### DISCUSSION TITLE
mapping GO terms to classifications level 2 (one rank down from the top- level) GO categories

### DISCUSSION THREAD

**Discussion by Ruiqi-CUB** [geneontology/helpdesk#431] [Q&A]
Dear Sir/madam,

I would like to map a list of GO terms (>1000, won't be able to do it manually) from different GO ranks to one classifications one rank down from the top- level GO categories, i.e. direct children categories of Biological Process, Cellular Component, and Molecular Function. Here is an example from [this paper](https://pubmed.ncbi.nlm.nih.gov/35600685/). Then I would like to do a bar graph like this. Which software/package should I use?

<img width="785" alt="Screen Shot 2023-02-27 at 11 15 12 AM" src="https://user-images.githubusercontent.com/46695842/221653682-31d85c08-d28f-4af7-be1c-e81f4f02f5e4.png">

Thanks a lot!

My GO list looks like this: [one GO term per line]

GO:0001505
GO:0006259
GO:0006403 ...

Ideal output: (Do not include those in parentheses)
 
GO:0001505               GO:0065007(level 2 GO terms)                 biological regulation  (level 2 GO term  description)        
GO:0006259              GO:0008152(level 2 GO terms 1/2)            metabolic process
GO:0006259              GO:0009987(level 2 GO terms 2/2)           cellular process         
 ...

Best Regards
Ruiqi

**Comment by suzialeksander** [comment]
So the largest concern with this is how useful this task is.  First of all, the level of the term doesn't really mean anything, it's just an artefact of how the ontology is constructed.  See http://geneontology.org/docs/faq/#how-can-i-calculate-the-level-of-a-go-term and the linked [Alterovitz et al.](https://academic.oup.com/nar/article/35/suppl_1/D322/1094033?login=false) 

A closely related but different concern is that the actual terms you'd get by just selecting children of the root terms may not even be used for annotations, or other concerns.  Take [GO:0005488 binding](http://amigo.geneontology.org/amigo/term/GO:0005488) for example- it's a direct child of GO:0003674
molecular_function, so at the 'level' you're asking about. However, the comments for that term are 

> Note that this term is in the subset of terms that **_should not be used for direct, manual gene product annotation_**. Please choose a more specific child term, or request a new one if no suitable term is available. For ligands that bind to signal transducing receptors, consider the molecular function term 'receptor binding ; GO:0005102' and its children.

Instead, you might consider using a curated subset (Slim): see http://geneontology.org/docs/go-subset-guide.  These contain a handpicked subgroup of terms in the specified aspect, but they are informative.  The [generic slim](http://geneontology.org/docs/download-ontology/#subsets) should be suitable for multi-organism files but there are a few Slims curated by specific MODs if you're only looking at one organism.

**Reply by Ruiqi-CUB** [reply]
Thanks a lot for the detailed reply! I saw lots of those figures in other papers and didn't even know they are wrong! 
I would like to show the overall trends in my organism but there are too many GO terms.  Is there any method you recommend!?  Thanks!

**Reply by Ruiqi-CUB** [reply]
Also,  you mentioned goslim_generic file.  What should I do after downloading the file?  Could I use it to map my GO terms to higher ranks?

**Reply by suzialeksander** [reply]
I think you would be interested in the web-based Slim Mapper tool.  It will take your list of gene products and sort them into pre-determined bins; these bins are the terms in the Slim. For example, there are only 23 terms ("bins") in the Yeast-specific Component slim.  Even if you load a list of all 6600+ yeast genes, they will only be sorted into those 23 bins or remain in an "unsortable" category where they are not annotated to a term in the slim, but they do have non-root annotations. The terms in the curated slims are selected in order to minimise the results in this unsortable category, as well as be informative and representative of the entire ontology.  You can use the pre-loaded Slims or load your own (for example, if you're only interested in one specific area of the GO that is not well-represented by the generic slim). The [Princeton tool](https://go.princeton.edu/cgi-bin/GOTermMapper) is compatible with several organisms, while the [SGD version](https://www.yeastgenome.org/goSlimMapper) is yeast-specific but they are essentially the same tool.  However, the help pages for each are equally helpful for understanding what the tool does and when you may want to use it.  See [Princeton's help page](https://go.princeton.edu/GOTermMapper/GOTermMapper_help.shtml#requiredOptions) and [SGD's help page](https://sites.google.com/view/yeastgenome-help/analyze-help/go-slim-mapper). 

The above tools are similar to but work differently than the GO Term Finder tool, also available at [Princeton GO Term Finder tool](https://go.princeton.edu/cgi-bin/GOTermFinder) and [SGD](https://www.yeastgenome.org/goTermFinder).  Instead of filing your genes into the preset bins, the  GO Term Finder will find common (shared) GO terms in the entire ontology.  In contrast to the Yeast slim mentioned above, there are 4043 terms in the component ontology that could be selected by this tool.  The GO Term Finder might be useful if the bins in a slim are too broad, but the results may not be as easily readable as they are with the Slim Mapper results. 

If you use the Slim Mapper or Term Finder (I use the SGD names but the Princeton names are similar), you don't need to download the slims.  The slim downloads tend to be more useful for tool developers or if you have another tool that works similar to the above Slim Mapper, and could then map your GO terms to the terms in the slim.  Note the GO Term Finder and GO Slim Mapper need genes/gene products, not GO terms.  Generally UniProt IDs will work in most tools, but you will need to check the Princeton (or SGD) docs to determine if your organism/IDs will be compatible with the tool.

**Reply by Ruiqi-CUB** [reply]
Thank you so much for the detailed explaination! Yeah, as you described, "take your list of gene products and sort them into pre-determined bins" is exactly what I want to do. 
However, I am working with a non-model bivalve species. I had a look at the tools you mentioned. It seems that there are only limit species options there and none of them are bivalves. Do you happen to know any similar tools that are designed for generic organisms?

**Comment by suzialeksander** [comment]
I believe you can use the Princeton tool's Advanced Options to load your own GAF, and therefore use Term Mapper with your organism. 

To obtain the GAF, I would recommend using QuickGO- basically, for a non-model organism, you will be more dependent on IEA annotations and [QuickGO has more of those loaded than AmiGO ](http://geneontology.org/docs/faq/#what-are-the-differences-between-the-data-available-in-amigo-and-those-on-quickgo).  You'll want to use the Taxon filter to display just your organism, so you'll be looking at a page similar to this:

https://www.ebi.ac.uk/QuickGO/annotations?taxonId=38949&taxonUsage=descendants

Use the Export button and select GAF. 

This will work if you have UniProt IDs for your genes; if you don't have them as UniProt IDs you may be able to use https://www.uniprot.org/id-mapping.

Once you have both a gene list (.txt is fine) and a GAF, use the Advanced Options as described in [the Princeton help page](https://go.princeton.edu/GOTermMapper/GOTermMapper_help.shtml).  So you'll fill 

> **Basic Options**
> 1. upload your gene ID list
> 2. choose your aspect
> 3. ignore
> 4. leave as Generic Ontology
> 
> **Advanced Options**
> 
> 2. upload the GAF obtained from QuickGO

Ignore the rest, unless you want to provide an email but I find that unnecessary as the tool is quite fast.

If you have quite a few genes in the "not annotated to a term in the slim, but they do have non-root annotations" list, you may consider running this sub-list through the GO Term Finder.  You'll need to load your that QuickGO GAF in that tool as well.

**Reply by Ruiqi-CUB** [reply]
Thanks a lot! Sorry I am still new to GO .I am exploring the transcriptome data and got some enriched GO terms in several comparisons, but there are two many of them. That's why I would like to "sort them into pre-determined bins".  Would you mind having a look at the following questions? Thanks a lot! 

1.  I have found ~80 GO terms on QuickGO in my species and successfully downloaded the GAF. Would the number of GOs be a concern? Are those GO terms the "pre-determined bins"? Here is a link to the species. 
[https://www.ebi.ac.uk/QuickGO/annotations?taxonId=561460&taxonUsage=descendants&downloadLimit=1000](https://www.ebi.ac.uk/QuickGO/annotations?taxonId=561460&taxonUsage=descendants&downloadLimit=1000)

2. I am still a bit confused by UniProt IDs for my genes. I have the eggnog mapper annotation results for my species, which has the following columns. Which column and which database would you recommend me to use for the mapping? Gene names? Why do we still need the mapping (see Q3)? 
`query_name(1)	seed_eggNOG_ortholog(2)	seed_ortholog_evalue(3)	seed_ortholog_score(4)	best_tax_level(5)	Preferred_name(6) GOs(7)	EC(8)	KEGG_ko(9)	KEGG_Pathway(10)	KEGG_Module(11)	KEGG_Reaction(12)	KEGG_rclass(13) BRITE(14)	KEGG_TC(15)	CAZy(16)	BiGG_Reaction(17)	taxonomic scope(18)	eggNOG OGs(19) best eggNOG OG(20)	COG(21) Functional cat.	eggNOG free text desc.(22)`

3.For the Princeton GO Term tool, should I upload the GO list that are enriched in the following format at 1B, then upload the GAF I downloaded from quickGO before? It seems that we don't need the UniProt IDs you mentioned?

```
Ada2/Gcn5/Ada3 transcription activator complex ; GO:0005671  
Arp2/3 protein complex ; GO:0005885 
```
<img width="682" alt="image" src="https://user-images.githubusercontent.com/46695842/222030756-123db07a-a96d-493c-8cd5-fa46d3f22468.png">

**Reply by suzialeksander** [reply]
Apologies for the delay @Ruiqi-CUB, I typed this up quite a while ago but looks like the post didn't go through. 

1. So the overall number of annotations, 82, is a bit low but it looks like your organism isn't yet well sequenced: https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?lvl=0&amp;id=561460 .  [These IEAs are based on submitted sequences](https://www.ebi.ac.uk/GOA/InterPro2GO) and will remain low until more of the sequence is submitted to NCBI or UniProt.   These 82 annotations are not mapped to slims, but as these are electronically assigned annotations based on traits like known domains, they may be to relatively higher-level terms than if they'd been manually curated from individual wet-lab experiments. 

2. I'll tag @cmungall about the eggnog results but there is an option to select eggNOG-> UniProtKB at https://www.uniprot.org/id-mapping.

3. For the GO Term Mapper, that 1B is if you have a custom subset (Slim) that you'd rather use instead of the generic.  For example, the yeast slim may have terms involving bud site selection and other terms specific to budding yeast, but not applicable to flies.  The generic slim should be compatible with all organisms, but may not have specifics like bud site- these may instead go in a generic slim bin about reproduction, if those terms are indeed related. 

   If you mean to sort your list of proteins into bins, you'll need to supply your list of proteins in Basic- step 1 and your organism's entire known annotation file (the GAF) in Advanced-step2.  

   Advanced-step1B is for loading your own slim, which if you're using the generic slim (and I would use the generic, at least until you have a better grasp of the tool and how it works) you can safely ignore that A-1B option. 

  The reason you'd need to convert your list of genes/proteins to UniProt IDs is the GAF you're supplying has UniProt IDs, and the tool does not have the ability or information needed to map IDs between the files if they are from different databases.
